src/libmw/src/node/CoinsViewFactory.cpp

```diff
@@ -0,0 +1,240 @@
+	 */
+	File(mmr::LeafSet::GetPath(chainDir, mmr_info.index))
+		.Write(leafset.bytes());
+	auto pLeafSet = mmr::LeafSet::Open(chainDir, mmr_info.index);
+	if (pLeafSet->Root() != pStateHeader->GetLeafsetRoot()) {
+		ThrowValidation(EConsensusError::MMR_MISMATCH);
+	}
+
+	/**
+	 * Build PruneList
+	 */
+	BitSet compact_bitset = mmr::MMRUtil::BuildCompactBitSet(pStateHeader->GetNumTXOs(), leafset);
+	File(mmr::PruneList::GetPath(chainDir, mmr_info.compact_index))
+		.Write(compact_bitset.bytes());
+	auto pPruneList = mmr::PruneList::Open(chainDir, mmr_info.compact_index);
+
+	/**
+	 * Build KernelMMR
+	 */
+	auto pBatch = pDBWrapper->CreateBatch();
+    auto pKernelMMR = BuildAndValidateKernelMMR(
+        pDBWrapper,
+		pBatch,
+		mmr_info,
+        pChain,
+        chainDir,
+        pStateHeader,
+        kernels
+    );
+	
+	/**
+	 * Build Output PMMR
+	 */
+	auto pOutputPMMR = BuildAndValidateOutputMMR(
+		pDBWrapper,
+		pBatch,
+		mmr_info,
+		chainDir,
+		pStateHeader,
+		utxos,
+		leafset,
+		pPruneList,
+		pruned_parent_hashes
+	);
+
+	/**
+	 * Validate block sums
+	 */
+    std::vector<Commitment> utxo_commitments;
+    std::transform(
+        utxos.cbegin(), utxos.cend(),
+        std::back_inserter(utxo_commitments),
+        [](const UTXO::CPtr& pUTXO) { return pUTXO->GetCommitment(); }
+    );
+
+	KernelSumValidator::ValidateState(
+		utxo_commitments,
+		kernels,
+		pStateHeader->GetKernelOffset()
+	);
+
+	// Add UTXOs to database
+	CoinDB coinDB(pDBWrapper.get(), pBatch.get());
+	coinDB.AddUTXOs(utxos);
+	pBatch->Commit();
+
+	return std::make_shared<mw::CoinsViewDB>(
+		pStateHeader,
+		pDBWrapper,
+		pLeafSet,
+		pKernelMMR,
+		pOutputPMMR
+	);
+}
+
+// TODO: Do we also want to validate peg-in/peg-out transactions beyond the horizon?
+// This will require us to save the hogex txs to disk, and never prune them.
+mmr::MMR::Ptr CoinsViewFactory::BuildAndValidateKernelMMR(
+    const std::shared_ptr<libmw::IDBWrapper>& pDBWrapper,
+	const std::unique_ptr<libmw::IDBBatch>& pBatch,
+	const MMRInfo& mmr_info,
+	const libmw::IChain::Ptr& pChain,
+    const FilePath& chainDir,
+    const mw::Header::CPtr& pStateHeader,
+    const std::vector<Kernel>& kernels)
+{
+    mmr::MMR::Ptr pMMR = std::make_shared<mmr::MMR>(
+		mmr::FileBackend::Open('K', chainDir, mmr_info.index, pDBWrapper, nullptr)
+	);
+
+	auto pChainIter = pChain->NewIterator();
+	assert(pChainIter->Valid());
+
+	std::vector<mmr::Leaf> leaves;
+	leaves.reserve(kernels.size());
+
+	uint64_t kernels_added = 0;
+    for (const Kernel& kernel : kernels)
+    {
+		if (!pChainIter->Valid()) {
+			ThrowValidation(EConsensusError::MMR_MISMATCH);
+		}
+
+        mmr::LeafIndex leaf_idx = pMMR->Add(kernel);
+		leaves.push_back(mmr::Leaf::Create(leaf_idx, kernel.Serialized()));
+
+		++kernels_added;
+
+        // We have to loop here because some blocks may not have any new kernels.
+        while (pChainIter->Valid() && kernels_added == pChainIter->GetHeader().pHeader->GetNumKernels())
```

Comment:
If at any point in the iteration kernels_added > pChainIter->GetHeader().pHeader->GetNumKernels() then that could pose a problem. None of the subsequent block headers in the chain would be checked.